src/pudl/extract/dbf.py

```diff
@@ -0,0 +1,301 @@
+        ...
+
+    def get_table_schema(self, table_name: str, year: int) -> TableSchema:
+        """Returns schema for a given table and a given year."""
+        ...
+
+    def load_table_dfs(self, table_name: str, years: list[int]) -> pd.DataFrame | None:
+        """Returns dataframe that contains data for a given table across given years."""
+        ...
+
+
+class FercFieldParser(FieldParser):
+    """A custom DBF parser to deal with bad FERC data types."""
+
+    def parseN(self, field, data: bytes) -> int | float | None:  # noqa: N802
+        """Augments the Numeric DBF parser to account for bad FERC data.
+
+        There are a small number of bad entries in the backlog of FERC Form 1
+        data. They take the form of leading/trailing zeroes or null characters
+        in supposedly numeric fields, and occasionally a naked '.'
+
+        Accordingly, this custom parser strips leading and trailing zeros and
+        null characters, and replaces a bare '.' character with zero, allowing
+        all these fields to be cast to numeric values.
+
+        Args:
+            field: The DBF field being parsed.
+            data: Binary data (bytes) read from the DBF file.
+        """  # noqa: D417
+        # Strip whitespace, null characters, and zeroes
+        data = data.strip().strip(b"*\x00").lstrip(b"0")
+        # Replace bare periods (which are non-numeric) with zero.
+        if data == b".":
+            data = b"0"
+        return super().parseN(field, data)
+
+
+DBF_TYPES = {
+    "C": sa.String,
+    "D": sa.Date,
+    "F": sa.Float,
+    "I": sa.Integer,
+    "L": sa.Boolean,
+    "M": sa.Text,  # 10 digit .DBT block number, stored as a string...
+    "N": sa.Float,
+    "T": sa.DateTime,
+    "0": sa.Integer,  # based on dbf2sqlite mapping
+    "B": "XXX",  # .DBT block number, binary string
+    "@": "XXX",  # Timestamp... Date = Julian Day, Time is in milliseconds?
+    "+": "XXX",  # Autoincrement (e.g. for IDs)
+    "O": "XXX",  # Double, 8 bytes
+    "G": "XXX",  # OLE 10 digit/byte number of a .DBT block, stored as string
+}
+"""dict: A mapping of DBF field types to SQLAlchemy Column types.
+
+This dictionary maps the strings which are used to denote field types in the DBF objects
+to the corresponding generic SQLAlchemy Column types: These definitions come from a
+combination of the dbfread example program dbf2sqlite and this DBF file format
+documentation page: http://www.dbase.com/KnowledgeBase/int/db7_file_fmt.htm
+
+Unmapped types left as 'XXX' which should result in an error if encountered.
+"""
+
+# TODO(rousik): circular dependency between FercFoxProDatastore and FoxProTable,
+# could be addressed through interfaces/ABCs
+
+
+# TODO(rousik): instead of using class-level constants, we could pass the params in the constructor, which should
+# allow us to instantiate these dataset-specific datastores in the extractor code.
+# That may make the manipulations little easier.
+class FercFoxProDatastore:
```

Comment:
This class **has** a Datastore, but it is not itself a Datastore, which I think is different from how we've managed the other data source specific Datastores, and might be confusing. What are the trade-offs / reasons for doing it this way rather than inheriting?

Reply:
As it is now, `FercFoxProDatastore` effectively wraps generic datastore and offers very different higher level interface for reading data and table schemas. As such, it doesn't really make sense to inherit, because we're redefining the offered functionality. If the naming is confusing, we could perhaps consider not naming it `Datastore`. How about `FercFoxProReader` or something like that?

Reply:
I think that non-Datastore naming would be less confusing, given the number of other Datastore subclasses we have flying around. Maybe `FercDbfReader`?